magda-io/magda
# Issue/1406 DGA Ckan redirection

### What this PR does

Redirect DGA URLs according to:
https://docs.google.com/spreadsheets/d/1nEyknNJj2bzv1sBltjGPL92j50hW9C5yUpcTYQnvzqM

How to test:
- Access URL: https://issue-1406.dev.magda.io/
- Can use URLs from the `DGA URL Sample` column of the google sheet above
  - Needs to replace the domain of the URL to `issue-1406.dev.magda.io` before copy to browser
- Currently, the target `ckan.data.gov.au` domain is not available yet. So you won't be able to see DGA side function --- as long as your browser is redirected, it's passed.
- If target function is located at Magda, you should make sure Madga's function works well after redirection.

You can also find some edging cases from unit tests.

For Code Review, some variation from the initial plan:
- As we don't have a good way to query dataset ID from a distribution ID, the code use `ckan resource id` to get ckan `package_id`. Then, use `package_id` to get `dataset_id`
- As aspect `organization-details` doesn't have a ckan `id` field, we query aspect `ckan-dataset`'s `organization` field and then get aspect `dataset-publisher` for magda side's publisher ID.
- There is an issue with typescript registry api client (api.ts see #1511). Because of that, we use `request` lib to request registry API directly.

Updated on 1st Aug 2018:

Test Links:
https://issue-1406.dev.magda.io/about
https://issue-1406.dev.magda.io/api/3/action/package_search?sort=extras_harvest_portal+asc
https://issue-1406.dev.magda.io/dataset/pg_skafsd0_f___00120141210_11a
https://issue-1406.dev.magda.io/dataset/8beb4387-ec03-46f9-8048-3ad76c0416c8
https://issue-1406.dev.magda.io/dataset/pg_skafsd0_f___00120141210_11a/resource/af618603-e529-4998-b977-e8751f291e6e
https://issue-1406.dev.magda.io/dataset/edit?q=xxx
https://issue-1406.dev.magda.io/dataset/new?q=23
https://issue-1406.dev.magda.io/fanstatic/ckanext-pdfview/:version:2017-02-15T10:30:47/css/pdf.css
https://issue-1406.dev.magda.io/geoserver/web/
https://issue-1406.dev.magda.io/group/business
https://issue-1406.dev.magda.io/organization
https://issue-1406.dev.magda.io/organization?q=ABS&sort=title+asc
https://issue-1406.dev.magda.io/organization/australianbureauofstatistics-geography
https://issue-1406.dev.magda.io/organization/760c24b1-3c3d-4ccb-8196-41530fcdebd5
https://issue-1406.dev.magda.io/showcase
https://issue-1406.dev.magda.io/storage/f/2013-05-12T173818/tmpMA2c8PConvict_records.txt
https://issue-1406.dev.magda.io/uploads/group/20140208-072514.709318bureauofmeteorologylogo.jpg
https://issue-1406.dev.magda.io/user/477pitt
https://issue-1406.dev.magda.io/vendor/leaflet/0.7.3/images/marker-shadow.png
https://issue-1406.dev.magda.io/stats#summary
https://issue-1406.dev.magda.io/datarequest
https://issue-1406.dev.magda.io/organization/act-government .  (Should see a not found page)
https://issue-1406.dev.magda.io/organization/datarequest/alpine-shire-council
https://issue-1406.dev.magda.io/organization/activity/alpine-shire-council/0
https://issue-1406.dev.magda.io/organization/about/alpine-shire-council
https://issue-1406.dev.magda.io/dataset/groups/wyndham-smart-bin-fill-level
https://issue-1406.dev.magda.io/dataset/activity/wyndham-smart-bin-fill-level
https://issue-1406.dev.magda.io/dataset/showcases/wyndham-smart-bin-fill-level
https://issue-1406.dev.magda.io/dataset/wyndham-smart-bin-fill-level/resource/15732b49-3e50-40ce-8dfd-0efed18661f4
https://issue-1406.dev.magda.io/dataset/08531201-ac9f-4f5f-bb7e-ac16b1da28b4/resource/15732b49-3e50-40ce-8dfd-0efed18661f4/download/sb_fill_lvel.json
https://issue-1406.dev.magda.io/dataset/certificate-transparency/resource/b718232a-bc8d-49c0-9c1f-33c31b57cd88/view/e70cce27-be4c-415a-99a2-1537ecc295ee

test not exist link (or any not exist link):
https://issue-1406.dev.magda.io/sdds/sdd/sdsds

If database, registry or any backend errors happend, user should see this page:
https://issue-1406.dev.magda.io/error?errorCode=500
(Probably hard to test this as hard to produce backend failure)

### Checklist

-   [x] I've linked this PR to an issue in ZenHub (or there is no issue) and dragged it to the _Pull Request_ column
-   [x] There are sufficient comments for my code to be understandable - and I realise reviewers will pull me up on it if not!
